skeleton/src/main/java/com/dropbox/kaiken/skeleton/skeleton/usermanagement/UserManager.kt

```diff
@@ -0,0 +1,119 @@
+package com.dropbox.kaiken.skeleton.skeleton.usermanagement
+
+import com.dropbox.kaiken.skeleton.skeleton.usermanagement.auth.ActiveUserManager
+import com.dropbox.kaiken.skeleton.skeleton.usermanagement.auth.SkeletonAuthInteractor
+import com.dropbox.kaiken.scoping.AppScope
+import com.squareup.anvil.annotations.ContributesBinding
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.drop
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.scan
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import javax.inject.Inject
+
+/**
+ * Manages the set of authenticated users for an app.
+ */
+interface UserManager {
+    /**
+     * Sets the given user ID to the active user. This returns true if the user ID is known, false otherwise.
+     */
+    suspend fun setActiveUser(newActiveUserId: String): Boolean
+
+    /**
+     * Returns the active user. If there is no active user, this will return null.
+     */
+    suspend fun getActiveUser(): User?
+
+    /**
+     * Provides a flow of the [UserState].  Observe this for changes related to users added, removed, and active users updated.
+     *
+     */
+    fun getUserState(): Flow<UserState>
+}
+
+@ExperimentalCoroutinesApi
+@ContributesBinding(AppScope::class)
+class RealUserManager @Inject constructor(
+        private val skeletonAuthInteractor: SkeletonAuthInteractor,
+        private val activeUserManager: ActiveUserManager,
+) : UserManager {
+
+    private val mutex = Mutex()
+    private val activeUserState = MutableStateFlow<String?>(null)
+
+    override suspend fun setActiveUser(newActiveUserId: String): Boolean {
+        return mutex.withLock {
+            val prevActiveUserId = activeUserManager.getActiveUserId()
+            if (newActiveUserId != prevActiveUserId && hasUser(newActiveUserId)) {
+                activeUserManager.setActiveUser(newActiveUserId)
+                activeUserState.emit(newActiveUserId)
+                true
+            } else {
+                false
+            }
+        }
+    }
+
+    private suspend fun hasUser(userId: String): Boolean = getUserState().first().users.any { account -> account.userId == userId }
+
+    override suspend fun getActiveUser(): User? = getUserState().first().users.firstOrNull { account -> account.isActiveUser }
+
+    override fun getUserState(): Flow<UserState> =
+            skeletonAuthInteractor.observeAuthChanges().combine(activeUserState) { usersInput, _ ->
```

Comment:
Should we default the first user added to active if none is currently set?

Reply:
follow up left a todo